### ISSUE TITLE
Add support for Half in EF Core

### ISSUE THREAD

**Issue by newmasterSG** [dotnet/efcore#35294]
Need to create full [half type](https://devblogs.microsoft.com/dotnet/introducing-the-half-type/) support for ef core. 

1.  Value converters, 
2.  Compiled models, 
3.  Migrations, 
4.  Mapping

**Comment by newmasterSG** [comment]
@bricelam Hi, could we discuss how to create full supporting half type for ef core? May be all familiar to float or decimal type?

**Comment by newmasterSG** [comment]
@roji Yeah, no problem, in this [pr](https://github.com/dotnet/efcore/pull/35193#issuecomment-2520682732) ef core decided to add whole supporting half type. As I see this, actually it will give for future developers get more flexibility which accuracy they what

**Comment by roji** [comment]
@newmasterSG so you don't have any concrete need yourself for Half support in EF, right? No database that I know of supports Half - that type is useful mainly for faster, lower-accuracy floating point calculations, which are generally less relevant in relational databases (a vector of Halfs might make more sense, but that's not what this is about).

**Comment by newmasterSG** [comment]
@roji Hm, that's make more sense, thank you for your explanation, so maybe that's better to close this issue, however I didn't fully understand, why team said that need to create separate issue for whole ef

**Comment by roji** [comment]
@newmasterSG note that I'm not saying we shouldn't merge the PR (or close this issue) - just trying to understand what the actual real-world usages would be of using Half in SQLite.

**Comment by bricelam** [comment]
To quote the .NET team:

> Computations in machine learning take advantage of the decreased precision to speed up calculations and cut down on storage space. So, this would probably be the primary use case for a Half.

**Comment by cincuranet** [comment]
Just to clarify for myself... Do we discuss here `Half` in EF Core or in `Microsoft.Data.Sqlite`? The issue title says EF Core, but reading in between the lines it looks like we're also talking about `Microsoft.Data.Sqlite`.

To reiterate current team's status:
* We're happy to take contribution of adding `Half` into `Microsoft.Data.Sqlite` (#30931). Majority believe (and yes, we had to massage @roji a bit ;)) that it falls into same category as `short`, `float`, or `TimeOnly`. You can go ahead with it @newmasterSG.
* Adding it into EF Core needs more understanding on our side why to do it "in the box". Instead of i.e. utilizing _value converter_ (given that virtually no database at the moment supports such datatype). Basically what the scenario is and how we can help making it more awesome by supporting `Half` there.

**Comment by roji** [comment]
@cincuranet just noting that if/when we add support to M.D.Sqlite, it makes sense to add Half support to the Sqlite EF provider as well - without needing any value converter (EF providers generally expose all types supported by the underlying ADO.NET provider). For other database, since there's no ADO.NET provider support, the only thing we can do is to add a value converter, which I don't think we should.

**Comment by bricelam** [comment]
To play devil’s advocate, what was the point of adding a Uri and IPAddress converter?

**Comment by ajcvickers** [comment]
> To play devil’s advocate, what was the point of adding a Uri and IPAddress converter?

That was my fault. Shay uses these as an example of where we did it wrong. 😆

**Comment by bricelam** [comment]
Then I believe Shay and I are in agreement. This becomes a provider-specific feature of M.D.SQLite and just for the sake of completeness.

**Comment by roji** [comment]
>> To play devil’s advocate, what was the point of adding a Uri and IPAddress converter?
>
> That was my fault. Shay uses these as an example of where we did it wrong. 😆

Just to be extra clear, I think adding EF-level value converters for arbitrary types (like IPAddress) is (much) more reasonable than adding ADO.NET support (when the type isn't natively supported) - EF is a much higher-level thing and goes way beyond "just exposing the database" (which ADO.NET is quite close to). That's also why I don't love the idea of adding Half support to M.D.Sqlite (it's not a standard, universally supported type like DateTime or decimal).

Even for EF I'd prefer for users to just specify value converters explicitly - especially where the conversion is as trivial as in this case - rather than EF doing magic which hides what's actually going on from the user. I wouldn't go so far as to say "we did it wrong" - it's not just my preference :)
